File: Graphs/801-is-graph-bipartite/is-graph-bipartite.cpp
class Solution {
public:
    bool dfs(vector<vector<int>>& graph, vector<int>& color, int cur, int n){
        color[cur] = n;

        for(int& i:graph[cur]){
            if(color[i] == -1){
                if(dfs(graph, color, i, !n) == false)
                    return false; 
            }
            else if(color[i] == n)
                return false;
        }

        return true;
    }

    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> color(n, -1);

        for(int i = 0; i < n; i++){
            if(color[i] == -1 && dfs(graph, color, i, 1) == false)
                return false;
        }

        return true;
    }
};

File: Graphs/801-is-graph-bipartite/is-graph-bipartite.py
class Solution:
    def dfs(self, graph, color, cur, n):
        color[cur] = n

        for neighbor in graph[cur]:
            if color[neighbor] == -1:
                if not self.dfs(graph, color, neighbor, 1 - n):
                    return False
            elif color[neighbor] == n:
                return False

        return True

    def isBipartite(self, graph):
        n = len(graph)
        color = [-1] * n

        for i in range(n):
            if color[i] == -1:
                if not self.dfs(graph, color, i, 1):
                    return False

        return True
